### PR TITLE
GH-1588 Conditional formatting for handson table.

### PR DESCRIPTION
**Pull request by mounicarachakonda**
Fixes GH-1588

Signed off by : Mounica Rachakonda <mounica.rachakonda@lexisnexis.com>

### CODE REVIEW

**Comment by GordonSmith** [comment]
@mounicarachakonda I think there are some changes in loader.js and gulpfile.js that will be needed also?

File: gulpfile.js [hunk 1]
```diff
@@ -30,8 +30,8 @@ const cfg = {
   prefix: "hpcc-viz"
 };
 
-const libs = ["d3", "c3", "colorbrewer", "dagre", "topojson", "d3-cloud", "font-awesome", "amcharts", "es6-promise", "amcharts.funnel", "amcharts.gauge", "amcharts.pie", "amcharts.radar", "amcharts.serial", "amcharts.xy", "amcharts.gantt", "amcharts.plugins.responsive", "simpleheat", "d3-hexbin", "d3-tip", "d3-bullet", "d3-sankey", "autoComplete"];
-const bundles = ["common", "layout", "api", "other", "chart", "form", "c3chart", "google", "amchart", "tree", "graph", "map", "composite", "marshaller"];  //  Order is important ---
+const libs = ["d3", "c3", "colorbrewer", "dagre", "topojson", "d3-cloud", "font-awesome", "amcharts", "es6-promise", "amcharts.funnel", "amcharts.gauge", "amcharts.pie", "amcharts.radar", "amcharts.serial", "amcharts.xy", "amcharts.gantt", "amcharts.plugins.responsive", "simpleheat", "d3-hexbin", "d3-tip", "d3-bullet", "d3-sankey", "autoComplete", "handsontable"];
+const bundles = ["common", "layout", "api", "other", "chart", "form", "c3chart", "google", "amchart", "tree", "graph", "map", "composite", "marshaller", "handson"];  //  Order is important ---
```
Comment (GordonSmith) [inline gulpfile.js]: Move "handson" to after "other" - the inter bundle dependency is important and it needs to be before the marshaller and composite.

File: src/handson/Table.css [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    background-color: #bce1fb;
+}
+
+.handson_Table .ht_master tr:nth-of-type(odd) > td {
+    background-color: #f3faff;
+}
+
+.handson_Table .ht_master tr:nth-of-type(even) > td {
+    //background-color: red;
+}
```
Comment (GordonSmith) [inline src/handson/Table.css]: Remove commented code.

File: test/handsonFactory.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
+"use strict";
+(function (root, factory) {
+    if (typeof define === "function" && define.amd) {
+        define([], factory);
+    } else {
+        root.test_handsonFactory = factory();
+    }
+}(this, function (callback) {
```
Comment (GordonSmith) [inline test/handsonFactory.js]: callback is invalid here

File: test/handsonFactory.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                    table.fields()[0].format(".2f");
+                    table.fields()[2].format(".6r");
+                    table.fields()[4].type("time").mask("%Y-%_m-%_d").format("%d %b '%y");
+                    callback(table);
+                });
+            }
+        },
+        }
+             return handsonFactory;
+
```
Comment (GordonSmith) [inline test/handsonFactory.js]: Indentation is wrong.

File: test/handsonFactory.js [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        define([], factory);
+    } else {
+        root.test_handsonFactory = factory();
+    }
+}(this, function (callback) {
+    var handsonFactory = {
+    	 HandsOnTable: {
+            simple: function (callback) {
+                require(["test/DataFactory", "src/handson/Table"], function (DataFactory, Table) {
+                    var table = new Table()
```
Comment (GordonSmith) [inline test/handsonFactory.js]: DataFactory is not used.

File: test/handsonFactory.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                });
+            },
+             widget: function (callback) {
+                require(["test/DataFactory", "src/handson/Table"], function (DataFactory, Table) {
+                    callback(new Table()
+                        .columns(DataFactory.Table.widget.columns)
+                        .data(DataFactory.Table.widget.data)
+                        //.minWidgetHeight(200)
+                        //.minWidgetWidth(200)
+                    );
```
Comment (GordonSmith) [inline test/handsonFactory.js]: remove commented code.

File: test/handsonFactory.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                });
+            },
+            large: function (callback) {
+                require(["test/DataFactory", "src/handson/Table"], function (DataFactory, Table) {
+                    callback(new Table()
+                    .columns(DataFactory.Table.large.columns)
+                    .data(DataFactory.Table.large.data)
```
Comment (GordonSmith) [inline test/handsonFactory.js]: indentation

File: test/handsonFactory.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+                    //.fixedHeader(true)
```
Comment (GordonSmith) [inline test/handsonFactory.js]: remove commented code.

File: src/handson/Table.js [hunk 9]
```diff
@@ -0,0 +1,351 @@
+"use strict";
+(function (root, factory) {
+    if (typeof define === "function" && define.amd) {
+        define(["d3", "handsontable", "../common/Widget", "../common/HTMLWidget", "../common/PropertyExt", "../common/Utility", "css!handsontable", 
+
+"css!./Table"], factory);
+    } else {
+        root.handson_Table = factory(root.d3, root.Handsontable, root.common_Widget, root.common_HTMLWidget, root.common_PropertyExt, root.common_Utility);
+    }
+}(this, function (d3, Handsontable, Widget, HTMLWidget, PropertyExt, Utility) {
```
Comment (GordonSmith) [inline src/handson/Table.js]: Is "Widget" used?

File: src/handson/Table.js [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+    function Column(owner) {
+        PropertyExt.call(this);
+        this._owner = owner;
+    }
+    Column.prototype = Object.create(PropertyExt.prototype);
+    Column.prototype.constructor = Column;
+    Column.prototype._class += " handson_Table.Column";
+
+    Column.prototype.publish("tableColumns", "", "set", "Only show column properties of the table",function () { return this._owner ? this._owner.columns() : []; }, { optional: true });
+    Column.prototype.publish("minRange", null, "number", "Minimum value of a handsontable", function () { return this._owner ? this._owner.columns() : []; }, { tags: ["Basic"], optional: true });
```
Comment (GordonSmith) [inline src/handson/Table.js]: ```function () { return this._owner ? this._owner.columns() : []; }``` is invalid here, should be ```null```;

File: src/handson/Table.js [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+    Column.prototype.publish("maxRange", null, "number", "Maximum value of a handsontable", function () { return this._owner ? this._owner.columns() : []; }, { tags: ["Basic"], optional: true });
```
Comment (GordonSmith) [inline src/handson/Table.js]: ```function () { return this._owner ? this._owner.columns() : []; }``` is invalid here, should be ```null```;

File: src/handson/Table.js [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+    Column.prototype.publish("belowMinRangeColor", "#ff0000", "html-color", "Series Color", function () { return this._owner ? this._owner.columns() : []; }, { optional: true });
```
Comment (GordonSmith) [inline src/handson/Table.js]: ```function () { return this._owner ? this._owner.columns() : []; }``` is invalid here, should be ```null```;

File: src/handson/Table.js [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+    Column.prototype.publish("aboveMaxRangeColor", "#0000ff", "html-color", "Series Color", function () { return this._owner ? this._owner.columns() : []; }, { optional: true });
```
Comment (GordonSmith) [inline src/handson/Table.js]: ```function () { return this._owner ? this._owner.columns() : []; }``` is invalid here, should be ```null```;

File: src/handson/Table.js [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+    Column.prototype.publish("rangeColor", "#00ff00", "html-color", "Series Color", function () { return this._owner ? this._owner.columns() : []; }, { optional: true });
```
Comment (GordonSmith) [inline src/handson/Table.js]: ```function () { return this._owner ? this._owner.columns() : []; }``` is invalid here, should be ```null```;

File: src/handson/Table.js [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+    
+    function Table() {
+        HTMLWidget.call(this);
+        this._tag = "div";
+
+        this._selectionBag = new Utility.Selection();
+        this._widgetCache = {};
+        this._widgetCache2 = {};
+    }
+//handson
```
Comment (GordonSmith) [inline src/handson/Table.js]: Invalid comment

File: src/handson/Table.js [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+    Table.prototype = Object.create(HTMLWidget.prototype);
+    Table.prototype.constructor = Table;
+    Table.prototype._class += " handson_Table";
+    Table.prototype.Column = Column;
+    Table.prototype.publish("mappings", [], "propertyArray", "Source Columns", null, { autoExpand: Column });
+    Table.prototype.publish("renderHtmlDataCells", false, "boolean", "enable or disable HTML within cells",null,{tags:["Private"]});
```
Comment (GordonSmith) [inline src/handson/Table.js]: Should not be "Private"

File: src/handson/Table.js [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+    Table.prototype.publish("fixedColumn", false, "boolean", "Enable or disable fixed first column", null, { tags: ["Private"] });
```
Comment (GordonSmith) [inline src/handson/Table.js]: Should not be Private

File: src/handson/Table.js [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+                    row = context._table.sortIndex[row][0];
+                }
+                if (context._widgetCache[row]) {
+                    return context.minWidgetHeight();
+                }
+                return undefined;
+            },
+            colWidths: function (col) {
+                if (context._widgetCache2[col]) {
+                    return context.minWidgetWidth();
```
Comment (GordonSmith) [inline src/handson/Table.js]: Need to test for "minWidgetWidth_exists" before using (as its optional)

File: src/handson/Table.js [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+
+            cells: function(row, col, prop) {
+                return {
+                    renderer: function (instance, td, row, col, prop, value, cellProperties) {
+                        if (row === 2 && col === 3) {
+                        }
+                        if (value instanceof WidgetReference) {
+                            var retVal = Handsontable.renderers.BaseRenderer.apply(this, arguments);
+
+                            var width = Math.max(context._table.getColWidth(col) || 50, context.minWidgetWidth()) - 8;
```
Comment (GordonSmith) [inline src/handson/Table.js]: Need to test for "minWidgetWidth_exists" before using (as its optional)

File: src/handson/Table.js [hunk 20]
```diff
@@ -0,0 +1,351 @@
...
+            //autoRowSize: {syncLimit: 10},
+            // autoColumnSize: {useHeaders: true},
+            stretchH: 'all',
+            currentRowClassName: 'currentRow',
+            rowHeights: function (row) {
+                if (context._table.sortIndex.length) {
+                    row = context._table.sortIndex[row][0];
+                }
+                if (context._widgetCache[row]) {
+                    return context.minWidgetHeight();
```
Comment (GordonSmith) [inline src/handson/Table.js]: Need to test for "minWidgetHeight_exists" before using (as its optional)

File: src/handson/Table.js [hunk 21]
```diff
@@ -0,0 +1,351 @@
...
+            cells: function(row, col, prop) {
+                return {
+                    renderer: function (instance, td, row, col, prop, value, cellProperties) {
+                        if (row === 2 && col === 3) {
+                        }
+                        if (value instanceof WidgetReference) {
+                            var retVal = Handsontable.renderers.BaseRenderer.apply(this, arguments);
+
+                            var width = Math.max(context._table.getColWidth(col) || 50, context.minWidgetWidth()) - 8;
+                            var height = Math.max(context._table.getRowHeight(row) || 16, context.minWidgetHeight());
```
Comment (GordonSmith) [inline src/handson/Table.js]: Need to test for "minWidgetHeight_exists" before using (as its optional)

File: src/handson/Table.js [hunk 22]
```diff
@@ -0,0 +1,351 @@
...
+            readOnlyCellClassName: "",
+            manualColumnResize: true,
+            sortIndicator: true,
+            wordWrap: false,
+            manualRowResize: true,
+            formulas : true,      
+            colHeaders: true,     
+            //autoColumnSize: {syncLimit: 10},
+            //autoRowSize: {syncLimit: 10},
+            // autoColumnSize: {useHeaders: true},
```
Comment (GordonSmith) [inline src/handson/Table.js]: Remove commented code

File: src/handson/Table.js [hunk 23]
```diff
@@ -0,0 +1,351 @@
...
+                        }, 0);
+                    }
+                }
+            },
+
+            cells: function(row, col, prop) {
+                return {
+                    renderer: function (instance, td, row, col, prop, value, cellProperties) {
+                        if (row === 2 && col === 3) {
+                        }
```
Comment (GordonSmith) [inline src/handson/Table.js]: Remove empty code

File: src/handson/Table.js [hunk 24]
```diff
@@ -0,0 +1,351 @@
...
+                        if (value instanceof WidgetReference) {
+                            var retVal = Handsontable.renderers.BaseRenderer.apply(this, arguments);
+
+                            var width = Math.max(context._table.getColWidth(col) || 50, context.minWidgetWidth()) - 8;
+                            var height = Math.max(context._table.getRowHeight(row) || 16, context.minWidgetHeight());
+                            if (!value.get(context).domNode) {
```
Comment (GordonSmith) [inline src/handson/Table.js]: Could ```value.get(context)``` return null?

File: src/handson/Table.js [hunk 25]
```diff
@@ -0,0 +1,351 @@
...
+
+cellProperties);
+                                break;
+                            case "boolean":
+                                Handsontable.renderers.CheckboxRenderer.call(this, instance, td, row, col, prop, value, 
+
+cellProperties);
+                                break;
+                            case "time":
+                                break;
```
Comment (GordonSmith) [inline src/handson/Table.js]: Should this fall through to default?

File: src/handson/Table.js [hunk 26]
```diff
@@ -0,0 +1,351 @@
...
+                            default:
+                                value = field.transform(value);
+                                textRender.call(this, instance, td, row, col, prop, value, cellProperties);
+                                break;
+                        }
+
+                        if (context._selectionBag.isSelected({ _id: row })) {
+                            td.style.color = "white";
+                            td.style.background = "#f48a00";
```
Comment (GordonSmith) [inline src/handson/Table.js]: How does the regular table handle selection colors?

File: src/handson/Table.js [hunk 27]
```diff
@@ -0,0 +1,351 @@
...
+
```
Comment (GordonSmith) [inline src/handson/Table.js]: remove blank line

File: src/handson/Table.js [hunk 28]
```diff
@@ -0,0 +1,351 @@
...
+                        } else {
+                            td.style.color = null;
+                            td.style.background = null;
+                        }
+                        if (row === context.data().length-1) {
+                               cellProperties.readOnly = false; // make cell read-only if it is first row or the text reads 'readOnly'
```
Comment (GordonSmith) [inline src/handson/Table.js]: Indentation + invalid comment - this makes the last row "not" readonly, is that correct?

File: src/handson/Table.js [hunk 29]
```diff
@@ -0,0 +1,351 @@
...
+                           if(col === context.columns().indexOf(field.__prop_label)){
+                                cellProperties.formulas = true;
+                                if(value === "sum("+field.__prop_label+")"){
+                                     var temp=0;
```
Comment (GordonSmith) [inline src/handson/Table.js]: indentation

File: src/handson/Table.js [hunk 30]
```diff
@@ -0,0 +1,351 @@
...
+                                                    break;
+                                                 }  
+                                        }
+                                    }  
+
+                                    td.textContent = temp;
+                                }
+                           }
+                        }else{
+                                cellProperties.readOnly = true;
```
Comment (GordonSmith) [inline src/handson/Table.js]: All cells are readonly except the last row?

File: src/handson/Table.js [hunk 31]
```diff
@@ -0,0 +1,351 @@
...
+                        }
+                        if(context.mappings().length>0){
+                            for (var k = 0; k < context.columns().length; k++) {
+                                if((context.mappings()[k]) !==undefined && ((context.mappings()[k]).tableColumns() === field.__prop_label)){
```
Comment (GordonSmith) [inline src/handson/Table.js]: Use field.label() not the private __prop_xxx

File: src/handson/Table.js [hunk 32]
```diff
@@ -0,0 +1,351 @@
...
+                            td.style.color = "white";
+                            td.style.background = "#f48a00";
+
+                        } else {
+                            td.style.color = null;
+                            td.style.background = null;
+                        }
+                        if (row === context.data().length-1) {
+                               cellProperties.readOnly = false; // make cell read-only if it is first row or the text reads 'readOnly'
+                           if(col === context.columns().indexOf(field.__prop_label)){
```
Comment (GordonSmith) [inline src/handson/Table.js]: Use field.label() not the private __prop_label member

File: src/handson/Table.js [hunk 33]
```diff
@@ -0,0 +1,351 @@
...
+                                cellProperties.formulas = true;
+                                if(value === "sum("+field.__prop_label+")"){
```
Comment (GordonSmith) [inline src/handson/Table.js]: Use field.label() not the private __prop_label member

File: src/handson/Table.js [hunk 34]
```diff
@@ -0,0 +1,351 @@
...
+
+    function Column(owner) {
+        PropertyExt.call(this);
+        this._owner = owner;
+    }
+    Column.prototype = Object.create(PropertyExt.prototype);
+    Column.prototype.constructor = Column;
+    Column.prototype._class += " handson_Table.Column";
+
+    Column.prototype.publish("tableColumns", "", "set", "Only show column properties of the table",function () { return this._owner ? this._owner.columns() : []; }, { optional: true });
```
Comment (GordonSmith) [inline src/handson/Table.js]: Should be "tableColumn" or "label" maybe?

File: src/handson/Table.js [hunk 35]
```diff
@@ -0,0 +1,351 @@
...
+                                    td.textContent = temp;
+                                }
+                           }
+                        }else{
+                                cellProperties.readOnly = true;
+                        }
+                        if(context.mappings().length>0){
+                            for (var k = 0; k < context.columns().length; k++) {
+                                if((context.mappings()[k]) !==undefined && ((context.mappings()[k]).tableColumns() === field.__prop_label)){
+                                        if (value < (context.mappings()[k]).minRange()){
```
Comment (GordonSmith) [inline src/handson/Table.js]: Need to test for ```minRange_exists()``` as its optional

File: src/handson/Table.js [hunk 36]
```diff
@@ -0,0 +1,351 @@
...
+                                          td.style.color = (context.mappings()[k]).belowMinRangeColor();
+                                        }else if(value > (context.mappings()[k]).maxRange()){
+                                             td.style.color = (context.mappings()[k]).aboveMaxRangeColor();
+                                        }else if(value >= (context.mappings()[k]).minRange() && value <= (context.mappings()[k]).maxRange()){
```
Comment (GordonSmith) [inline src/handson/Table.js]: Need to test for ```minRange_exists()``` as its optional

File: src/handson/Table.js [hunk 37]
```diff
@@ -0,0 +1,351 @@
...
+                           }
+                        }else{
+                                cellProperties.readOnly = true;
+                        }
+                        if(context.mappings().length>0){
+                            for (var k = 0; k < context.columns().length; k++) {
+                                if((context.mappings()[k]) !==undefined && ((context.mappings()[k]).tableColumns() === field.__prop_label)){
+                                        if (value < (context.mappings()[k]).minRange()){
+                                          td.style.color = (context.mappings()[k]).belowMinRangeColor();
+                                        }else if(value > (context.mappings()[k]).maxRange()){
```
Comment (GordonSmith) [inline src/handson/Table.js]: Need to test for maxRange_exists() as its optional

File: src/handson/Table.js [hunk 38]
```diff
@@ -0,0 +1,351 @@
...
+                                }
+                           }
+                        }else{
+                                cellProperties.readOnly = true;
+                        }
+                        if(context.mappings().length>0){
+                            for (var k = 0; k < context.columns().length; k++) {
+                                if((context.mappings()[k]) !==undefined && ((context.mappings()[k]).tableColumns() === field.__prop_label)){
+                                        if (value < (context.mappings()[k]).minRange()){
+                                          td.style.color = (context.mappings()[k]).belowMinRangeColor();
```
Comment (GordonSmith) [inline src/handson/Table.js]: Need to test for ```belowMinRangeColor()``` as its optional

File: src/handson/Table.js [hunk 39]
```diff
@@ -0,0 +1,351 @@
...
+                                        }else if(value > (context.mappings()[k]).maxRange()){
+                                             td.style.color = (context.mappings()[k]).aboveMaxRangeColor();
```
Comment (GordonSmith) [inline src/handson/Table.js]: Need to test for ```aboveMaxRangeColor()``` as its optional

File: src/handson/Table.js [hunk 40]
```diff
@@ -0,0 +1,351 @@
...
+    };
+
+    Table.prototype.update = function (domNode, element) {
+        HTMLWidget.prototype.update.apply(this, arguments);
+        var context = this;
+        var settings = {};
+        settings.colHeaders = context.columns();
+        settings.columns = this.fields().map(function (field) {
+            return {
+                //readOnly: true,
```
Comment (GordonSmith) [inline src/handson/Table.js]: remove commented code

File: src/handson/Table.js [hunk 41]
```diff
@@ -0,0 +1,351 @@
...
+            };
+        });
+        settings.fixedColumnsLeft = this.fixedColumn() ? 1 : 0;
+        settings.width = this.width();
+        settings.height = this.height();
+        settings.data = this.data().map(function (row, rowIdx) {
+            return row.map(function (cell, colIdx) {
+                if (cell instanceof Widget) {
+                    var retVal = new WidgetReference(rowIdx, colIdx);
+                    if (retVal.get(context).widget !== cell) {
```
Comment (GordonSmith) [inline src/handson/Table.js]: can this ever return null/undefined?

File: src/handson/Table.js [hunk 42]
```diff
@@ -0,0 +1,351 @@
...
+            this._table.sort(this._table.sortColumn, this._table.sortOrder);
+        }
+    };
+
+    Table.prototype.exit = function (domNode, element) {
+        HTMLWidget.prototype.exit.apply(this, arguments);
+    };
+
+    Table.prototype._createSelectionObject = function (i) {
+    //    var context = this;
```
Comment (GordonSmith) [inline src/handson/Table.js]: remove commented code

File: src/handson/Table.js [hunk 43]
```diff
@@ -0,0 +1,351 @@
...
+            var selObj = this._createSelectionObject(i);
+            this._selectionBag.click(selObj, { ctrlKey: this._selectionBag.isSelected(selObj) });
+            this._selectionPrevClick = d;
+        }
+    };
+
+    Table.prototype.renderInputs = function(d) {
+
+        alert("call render input");
+    };
```
Comment (GordonSmith) [inline src/handson/Table.js]: remove "alert" - is whole method needed?

File: src/handson/Table.js (hunk 9 above)
Comment (mounicarachakonda) [inline src/handson/Table.js]: Yes, It is used to check the cell instantiation with widget instance.

File: src/handson/Table.js (hunk 17 above)
Comment (mounicarachakonda) [inline src/handson/Table.js]: I have changed it from private to basic. Please let me know what is the exact difference between them here.

File: src/handson/Table.js [hunk 44]
```diff
@@ -0,0 +1,336 @@
...
+    function Column(owner) {
+        PropertyExt.call(this);
+        this._owner = owner;
+    }
+    Column.prototype = Object.create(PropertyExt.prototype);
+    Column.prototype.constructor = Column;
+    Column.prototype._class += " handson_Table.Column";
+
+    Column.prototype.publish("label", "", "set", "Only show column properties of the table", function() { return this._owner ? this._owner.columns() : [];}, {optional: true });
+    Column.prototype.publish("minRange", null, "number", "Minimum value of a handsontable", function() { return this._owner ? this._owner.columns() : [];});
```
Comment (GordonSmith) [inline src/handson/Table.js]: ```function () { return this._owner ? this._owner.columns() : []; }``` is invalid here (this param is only used for "set" PP types) - it should be null.

File: src/handson/Table.js [hunk 45]
```diff
@@ -0,0 +1,336 @@
...
+    Column.prototype.publish("maxRange", null, "number", "Maximum value of a handsontable", function() { return this._owner ? this._owner.columns() : [];});
```
Comment (GordonSmith) [inline src/handson/Table.js]: ```function () { return this._owner ? this._owner.columns() : []; }``` is invalid here (this param is only used for "set" PP types) - it should be null.

File: src/handson/Table.js [hunk 46]
```diff
@@ -0,0 +1,336 @@
...
+    Column.prototype.publish("belowMinRangeColor", "#ff0000", "html-color", "Series Color", function() { return this._owner ? this._owner.columns() : [];});
```
Comment (GordonSmith) [inline src/handson/Table.js]: ```function () { return this._owner ? this._owner.columns() : []; }``` is invalid here (this param is only used for "set" PP types) - it should be null.

File: src/handson/Table.js [hunk 47]
```diff
@@ -0,0 +1,336 @@
...
+    Column.prototype.publish("aboveMaxRangeColor", "#0000ff", "html-color", "Series Color", function() { return this._owner ? this._owner.columns() : [];});
```
Comment (GordonSmith) [inline src/handson/Table.js]: ```function () { return this._owner ? this._owner.columns() : []; }``` is invalid here (this param is only used for "set" PP types) - it should be null.

File: src/handson/Table.js [hunk 48]
```diff
@@ -0,0 +1,336 @@
...
+    Column.prototype.publish("rangeColor", "#00ff00", "html-color", "Series Color", function() { return this._owner ? this._owner.columns() : [];});
```
Comment (GordonSmith) [inline src/handson/Table.js]: ```function () { return this._owner ? this._owner.columns() : []; }``` is invalid here (this param is only used for "set" PP types) - it should be null.

File: src/handson/Table.js [hunk 49]
```diff
@@ -0,0 +1,336 @@
...
+
+    function Column(owner) {
+        PropertyExt.call(this);
+        this._owner = owner;
+    }
+    Column.prototype = Object.create(PropertyExt.prototype);
+    Column.prototype.constructor = Column;
+    Column.prototype._class += " handson_Table.Column";
+
+    Column.prototype.publish("label", "", "set", "Only show column properties of the table", function() { return this._owner ? this._owner.columns() : [];}, {optional: true });
```
Comment (GordonSmith) [inline src/handson/Table.js]: Can label be optional?

File: src/handson/Table.js [hunk 50]
```diff
@@ -0,0 +1,336 @@
...
+                            }
+                            Handsontable.Dom.empty(td);
+                            td.appendChild(retVal.get(context).domNode);
+                            return retVal;
+                        }
+
+                        var field = context.fields()[col];
+                        var textRender = context.renderHtmlDataCells() ? Handsontable.renderers.HtmlRenderer :
+
+                            Handsontable.renderers.TextRenderer;
```
Comment (GordonSmith) [inline src/handson/Table.js]: indentation and line break

File: src/handson/Table.js [hunk 51]
```diff
@@ -0,0 +1,336 @@
...
+
+                            var width = Math.max(context._table.getColWidth(col) || 50, context.minWidgetWidth()) - 8;
+                            var height = Math.max(context._table.getRowHeight(row) || 16, context.minWidgetHeight());
+                            if (retVal.get(context) && !value.get(context).domNode) {
+                                retVal.get(context).domNode = document.createElement('DIV');
+                                retVal.get(context).widget
+                                    .width(width)
+                                    .height(height)
+                                    .target(retVal.get(context).domNode)
+
```
Comment (GordonSmith) [inline src/handson/Table.js]: remove blank line

File: src/handson/Table.js [hunk 52]
```diff
@@ -0,0 +1,336 @@
...
+                            Handsontable.renderers.TextRenderer;
+                        switch (field.type()) {
+                            case "string":
+                                value = field.transform(value);
+                                textRender.call(this, instance, td, row, col, prop, value, cellProperties);
+                                break;
+                            case "number":
+                                Handsontable.renderers.NumericRenderer.call(this, instance, td, row, col, prop, value,
+
+                                    cellProperties);
```
Comment (GordonSmith) [inline src/handson/Table.js]: Remove line break(s) and indentation is wrong.

File: src/handson/Table.js [hunk 53]
```diff
@@ -0,0 +1,336 @@
...
+                                break;
+                            case "boolean":
+                                Handsontable.renderers.CheckboxRenderer.call(this, instance, td, row, col, prop, value,
+
+                                    cellProperties);
```
Comment (GordonSmith) [inline src/handson/Table.js]: Remove line break(s) and indentation is wrong.

File: src/handson/Table.js [hunk 54]
```diff
@@ -0,0 +1,336 @@
...
+                                value = field.transform(value);
+                                textRender.call(this, instance, td, row, col, prop, value, cellProperties);
+                                break;
+                        }
+                        if (context._selectionBag.isSelected({
+                                _id: row
+                            })) {
+                            td.style.color = "white";
+                            td.style.background = "#f48a00";
+
```
Comment (GordonSmith) [inline src/handson/Table.js]: remove blank line

**Comment by mounicarachakonda** [comment]
@GordonSmith , Please review.

File: src/handson/Table.js (hunk 30 above)
Comment (mounicarachakonda) [inline src/handson/Table.js]: Yes, In last row we can write the formula and it calculates the grand total of the column. sum(column name) gives the total of the column.

File: src/handson/Table.js [hunk 55]
```diff
@@ -0,0 +1,326 @@
...
+        this._selectionBag = new Utility.Selection();
+        this._widgetCache = {};
+        this._widgetCache2 = {};
+    }
+
+    Table.prototype = Object.create(HTMLWidget.prototype);
+    Table.prototype.constructor = Table;
+    Table.prototype._class += " handson_Table";
+    Table.prototype.Column = Column;
+    Table.prototype.publish("mappings", [], "propertyArray", "Source Columns", null, { autoExpand: Column });
```
Comment (GordonSmith) [inline src/handson/Table.js]: Rename to "columnFormatting" or similar?

**Comment by GordonSmith** [comment]
http://rawgit.com/mounicarachakonda/Visualization/handson/demos/dermatology.html?src/handson/HandsOnTable.simple
* Has blank line at the bottom - which I assume is for formula?  It should be optional?

**Comment by GordonSmith** [comment]
http://rawgit.com/mounicarachakonda/Visualization/handson/demos/dermatology.html?src/handson/HandsOnTable.widget
* Embedded widgets are not being displayed

(The original POC did:  http://rawgit.com/GordonSmith/Visualization/HANDS_ON_TABLE/demos/dermatology.html?src/other/HandsOnTable.widget)

**Pull request merged**